src/overload_ifelse.jl

```diff
@@ -0,0 +1,41 @@
+const SimpleTracer = Union{ConnectivityTracer,GradientTracer,HessianTracer}
+
+function Base.ifelse(::T, x, y) where {T<:SimpleTracer}
+    size(x) != size(y) &&
+        throw(DimensionMismatch("Outputs sizes of `ifelse` arguments don't match in size."))
```

Comment:
Why is this necessary? They don't have to both be arrays, they could be scalars or god knows what else

Reply:
We only support functions with scalar and array outputs and 
```Julia
julia> size(1) == size(42.0)
true
```

Reply:
We could also just let it error on the broadcast of `output_union`.

Reply:
We could let this error on the broadcast, but there are some edge cases where this could go wrong. 

Assume we have an (admittedly type unstable) function `ifelse(cond, expr1, expr2)`. The expression `expr1` computes a scalar `HessianTracer`, `expr2` an array of `HessianTracer`s. Simply broadcasting `expr1` and `expr2` would put the Hessian entries of `expr1` into all outputs of `expr2`.

This is why I added the size check `size(expr1) == size(expr2)`.

Reply:
> We only support functions with scalar and array outputs

That's true for the global function whose Jacobian we analyze. But inside of it there may be functions with arbitrary inputs and outputs:

```julia
struct MyStruct
    a
    b
end

function f(x::Vector)
    s = MyStruct(x[1], x[2])
    return s.a + s.b
end
```

I think the error for a missing broadcast of `size` would be more confusing.

If you want to avoid the situation `ifelse(b, x::T1, y::T2)`, we can define only `ifelse(b, x::T, y::T)`

Reply:
Good point with the structs and potentially missing broadcasts (isn't it a requirement of the `AbstractArray` interface though?).
However, if `expr1` and `expr2` are structs, we have no way to compute an `output_union` either. 

> If you want to avoid the situation ifelse(b, x::T1, y::T2), we can define only ifelse(b, x::T, y::T)

We can't do this since `ifelse(cond, ::AbstractTracer, ::Float64)` is completely valid and a common occurrence (e.g. `relu`).

Types also aren't a solution for my problem in case `expr1` and `expr2` are both arrays of different sizes but can broadcast.

Reply:
Sorry by broadcast I meant dispatch

Reply:
I'm just saying that the `size` check is not useful, if we have arrays the error will happen when trying to broadcast (not dispatch :sunglasses:) them together, and if we have weird objects it will just obscure the actual method that we need to implement

Reply:
> if we have arrays the error will happen when trying to broadcast (not dispatch 😎) them together

This isn't necessarily true, an `expr1` of size `(2, 3, 4)` will happily broadcast with an `expr2` of size `(2, 3)`.

Reply:
but then that's the behavior we want

Reply:
for arrays of numbers and for arrays of tracers

Reply:
No, we don't want arrays of tracers of size `(2, 3, 4)` to broadcast with arrays of tracers of size `(2, 3)`.

Reply:
why not?

Reply:
Using your example:
```Julia
julia> using SparseConnectivityTracer

julia> foo(x) = reshape(x[1:12], (2, 3, 2));

julia> bar(x) = reshape(x[13:18], (2, 3, 1));

julia> jacobian_pattern(foo, rand(18))
12×18 SparseArrays.SparseMatrixCSC{Bool, Int64} with 12 stored entries:
⎡⠑⢄⠀⠀⠀⠀⠀⠀⠀⎤
⎢⠀⠀⠑⢄⠀⠀⠀⠀⠀⎥
⎣⠀⠀⠀⠀⠑⢄⠀⠀⠀⎦

julia> jacobian_pattern(bar, rand(18))
6×18 SparseArrays.SparseMatrixCSC{Bool, Int64} with 6 stored entries:
⎡⠀⠀⠀⠀⠀⠀⠑⡀⠀⎤
⎣⠀⠀⠀⠀⠀⠀⠀⠈⢄⎦

julia> jacobian_pattern(x -> ifelse(x[1], foo(x), bar(x)), rand(18))  # <--- with size check
ERROR: DimensionMismatch: Outputs sizes of `ifelse` arguments don't match in size.
Stacktrace:
 [1] ifelse
   @ ~/Developer/SparseConnectivityTracer.jl/src/overload_ifelse.jl:2 [inlined]
 [2] (::var"#5#6")(x::Vector{SparseConnectivityTracer.GradientTracer{BitSet}})
   @ Main ./REPL[6]:1
 [3] trace_function(::Type{SparseConnectivityTracer.GradientTracer{BitSet}}, f::var"#5#6", x::Vector{Float64})
   @ SparseConnectivityTracer ~/Developer/SparseConnectivityTracer.jl/src/pattern.jl:32
 [4] jacobian_pattern(f::Function, x::Vector{Float64}, ::Type{BitSet})
   @ SparseConnectivityTracer ~/Developer/SparseConnectivityTracer.jl/src/pattern.jl:201
 [5] top-level scope
   @ REPL[6]:1

julia> jacobian_pattern(x -> ifelse(x[1], foo(x), bar(x)), rand(18))   # <--- without size check
12×18 SparseArrays.SparseMatrixCSC{Bool, Int64} with 24 stored entries:
⎡⠑⢄⠀⠀⠀⠀⠑⢄⠀⎤
⎢⠀⠀⠑⢄⠀⠀⢄⠀⠑⎥
⎣⠀⠀⠀⠀⠑⢄⠀⠑⢄⎦
```

Reply:
It is not immediately clear to me why this Jacobian pattern with "24 stored entries" for 18 inputs is correct.

Reply:
- The first 12 columns of the Jacobian correspond to `x[1:12]`. These components are used only once in the output
- The last 6 columns of the Jacobian correspond to `x[13:18]`. These components are used twice in the output because they are duplicated by broadcasting over the third dimension.

```julia
julia> using ForwardDiff, SparseArrays

julia> function f(x)
           y = reshape(x[1:12], (2, 3, 2))
           z = reshape(x[13:18], (2, 3, 1))
           return y .+ z
       end
f (generic function with 1 method)

julia> ForwardDiff.jacobian(f, rand(18)) |> sparse
12×18 SparseMatrixCSC{Float64, Int64} with 24 stored entries:
⎡⠑⢄⠀⠀⠀⠀⠑⢄⠀⎤
⎢⠀⠀⠑⢄⠀⠀⢄⠀⠑⎥
⎣⠀⠀⠀⠀⠑⢄⠀⠑⢄⎦
```

Reply:
Basically it's legal in Julia to do this so it should be legal in SCT

Reply:
> It is not immediately clear to me why this Jacobian pattern with "24 stored entries" for 18 inputs is correct.

If the Jacobians could only have as many nonzero entries as there are inputs, life would be boring ;)

Reply:
> Exactly. But why is this mathematically correct? Why do we want this behavior?


It's correct because it is what Julia does. The following functions are equivalent

```julia
julia> function f(x)
           y = reshape(x[1:12], (2, 3, 2))
           z = reshape(x[13:18], (2, 3, 1))
           return y .+ z
       end
f (generic function with 1 method)

julia> function g(x)
           y = reshape(x[1:12], (2, 3, 2))
           z = reshape(x[13:18], (2, 3))
           return y .+ cat(z, z; dims=3)
       end
g (generic function with 1 method)

julia> x = rand(18);

julia> f(x) == g(x)
true
```

> Even from a pure "ConnectivityTracer perspective", bar(x) = reshape(x[13:18], (2, 3, 1)) never even "influenced" the outputs at indices 7 to 12.

This is wrong, as evidenced by the ForwardDiff Jacobian (if there is a nonzero in the Jacobian, there is connectivity.

Reply:
> My take is: if m changes depending on the control flow, we throw an error. This is what the size check does.

Your definition of control flow is a bit extreme. By that token, we should error on the following function since `3` is "control flow":

```julia
h(x) = fill(x[1], 3)
```

Reply:
And if you need one more argument, this kind of implicit broadcasting over singleton dimensions is heavily used in deep learning

Reply:
> It's correct because it is what Julia does. 

Julia doesn't do sparsity detection. The question is about the mathematically correct way to do sparsity detection.

> The following functions are equivalent

I agree with your examples, but they are **not** in the context of `ifelse`.
Julia does **not** broadcast the two branches of an `ifelse` function.

> as evidenced by the ForwardDiff Jacobian (if there is a nonzero in the Jacobian, there is connectivity.

Once again, ForwardDiff does not broadcast the two branches of an `ifelse` function.

Reply:
The code in question is:
```Julia
function Base.ifelse(::AbstractTracer, result_of_some_computation, result_of_other_computation)
    size(result_of_some_computation) != size(result_of_other_computation) &&
        throw(DimensionMismatch("Outputs sizes of `ifelse` arguments don't match in size."))
    return compute_union_of_tracers_in_results(result_of_some_computation, result_of_other_computation)
end

Reply:
It's not inherently wrong for `ifelse` to choose between two objects of different sizes. Typically if there's a broadcast afterwards that is compatible with both

Reply:
For a less contrived example:

```julia
julia> g(x) = x .+ ifelse(x[1] > 0, x, 1)
g (generic function with 1 method)

julia> g([3, 4])
2-element Vector{Int64}:
 6
 8

julia> g([-3, 4])
2-element Vector{Int64}:
 -2
  5
```

Reply:
It's not blocking for me so if it's really important for you, go ahead. But you'll be forbidding perfectly legal Julia code like the one above, for which both global and local tracing make perfect sense.